pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go

```diff
@@ -0,0 +1,158 @@
+func (d *azureWIAuthProvider) Enabled() bool {
+	return true
+}
+
+// Provide returns the credentials for a specificed artifact.
+// Uses Azure Workload Identity to retrieve an AAD access token which can be
+// exchanged for a valid ACR refresh token for login.
+func (d *azureWIAuthProvider) Provide(artifact string) (provider.AuthConfig, error) {
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+
+	// parse the artifact reference string to extract the registry host name
+	artifactHostName, err := provider.GetRegistryHostName(artifact)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// retrieve an AAD Access token
+	aadToken, err := getAADAccessToken(tenantID, AADResource)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// send a challenge to the login server
+	directive, err := receiveChallengeFromLoginServer(artifactHostName, "https")
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// use challenge directive and AAD token to exchange for a registry token
+	refreshToken, err := performTokenExchange(artifactHostName, directive, tenantID, aadToken)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	authConfig := provider.AuthConfig{
+		Username: dockerTokenLoginUsernameGUID,
+		Password: refreshToken,
+		Provider: d,
+	}
+	return authConfig, nil
+}
+
+// Source: https://github.com/Azure/azure-workload-identity/blob/d126293e3c7c669378b225ad1b1f29cf6af4e56d/examples/msal-go/token_credential.go#L25
+func getAADAccessToken(tenantID, resource string) (string, error) {
+	// Azure AD Workload Identity webhook will inject the following env vars:
+	// 	AZURE_CLIENT_ID with the clientID set in the service account annotation
+	// 	AZURE_TENANT_ID with the tenantID set in the service account annotation. If not defined, then
+	// 	the tenantID provided via azure-wi-webhook-config for the webhook will be used.
+	// 	AZURE_FEDERATED_TOKEN_FILE is the service account token path
+	// 	AZURE_AUTHORITY_HOST is the AAD authority hostname
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+	tokenFilePath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+	authorityHost := os.Getenv("AZURE_AUTHORITY_HOST")
+
+	// read the service account token from the filesystem
+	signedAssertion, err := readJWTFromFS(tokenFilePath)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read service account token")
+	}
+	cred, err := confidential.NewCredFromAssertion(signedAssertion)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create confidential creds")
+	}
+
+	// create the confidential client to request an AAD token
+	confidentialClientApp, err := confidential.New(
+		clientID,
+		cred,
+		confidential.WithAuthority(fmt.Sprintf("%s%s/oauth2/token", authorityHost, tenantID)))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create confidential client app")
+	}
+
+	// .default needs to be added to the scope
+	if !strings.HasSuffix(resource, ".default") {
```

Comment:
Isn't this resource static in our case? Do you need this to build for every call?

Reply:
Yes you're right. In fact, what we can actually do is frontload the getAADToken operation to happen upon auth provider registration on pod startup since the AADtoken is not registry host name specific in this scenario.